### ISSUE TITLE
RFC: support for non-western systems

### ISSUE THREAD

**Issue by bmcfee** [librosa/librosa#641]
#### Description
This issue popped up during poster discussions at ismir2017 (tagging @vrangasayee): librosa's support for non-western (eg, Indian) scales is currently non-existent, but it would be easy to add.

Things that would be useful to have and trivial to implement

- Unit conversion functions: `hz_to_*`, `note_to_*`, etc
- Visualization decorations, `specshow(y_axis='...')`

More fundamentally, it would be nice to have support built in for `chroma`, and if possible, down to `cqt` basis filters.  The latter might take some refactoring, but if it works out, we should also be able to support non-equal-temperament scales.  (Consider these as stretch goals.)

**Comment by vrangasayee** [comment]
This would be really useful for a lot of non-western scales including Indian and other systems. Would be glad to collaborate to make this happen.

**Comment by bmcfee** [comment]
@vrangasayee @stefan-balke great!  I'm a total newbie at this, but is [this wiki page](https://en.wikipedia.org/wiki/Svara) a reasonable starting point?  Is there a better standard reference for notation?

As I mentioned above, the first thing to do is get the unit converters specified.  Do you know if there's a standard notation for svara including octave height and fractional notes?  Like, does `Sa3+35` make sense the way that `C3+35` does?  If so, then the converters should be pretty trivial to code up with `note_to_svara` as the core converter, from which we can derive `hz_to_svara`, `midi_to_svara`, etc.  

After that, the viz ticker decorations are pretty trivial, and then we can worry about chroma-like features, and then extending into (say) Chinese systems.  (Side note: any recommendations for who to consult on implementing those?)

**Comment by vrangasayee** [comment]
That looks reasonable. I have other texts that provide a better coverage
especially of the fractional notes. Let me try to scan and upload those (or
find online references).

**Comment by vrangasayee** [comment]
Here's a authoritative text on the 22 sruthi which also covers the 12 and
16 sruthis as ratios.

**Comment by stefan-balke** [comment]
Hm, I think your attachment got stuck somewhere. Can you share a dropbox link?

**Comment by vrangasayee** [comment]
[srutis.pdf](https://github.com/librosa/librosa/files/1466980/srutis.pdf)
Uploaded to github. This should work.

**Comment by bmcfee** [comment]
Thanks for that!  It's gonna take me a bit to digest, but a few questions immediately come to mind (apologies in advance if I'm being excessively ignorant):

1. It seems to me that this describes 3 systems, dividing into 22, 16, and 12 tones, with the latter two being subsets of the first.  Is that correct?  Can we get away with only implementing the 22?
2. Is there a [scientific pitch notation](https://en.wikipedia.org/wiki/Scientific_pitch_notation) standard that we can use here?  Is there a canonical relation from octave number to absolute frequency (eg `Sa1 ~= 32.67Hz`)?
3. We should support both Carnatic and Hindustani namings, right?  Any thoughts on how to implement that?
4. Are these always defined by just intonation, or can we define them by cent deviations instead?  The latter would be considerably simpler for tying into CQT / chroma representations, and also more consistent with the rest of the library.

**Comment by lostanlen** [comment]
ragaDB: a JSON graphical ontology of 163 Ragas and their parameters
https://github.com/shockmonger/ragaDB

**Comment by bmcfee** [comment]
Pinging this thread again -- we definitely have not lost interest in this, but I do think it needs the attention of an expert in these systems to make sure we implement it properly.

From my (very, very novice) understanding, Hindustani should be easy enough, but Carnatic would be trickier or need additional parameters to disambiguate certain classes depending on context.

I suppose the same thing occurs in western notes, and we've generally been sidestepping by calling all pitch classes outside of `C:maj` as sharp.  This makes me think we could expand the `_to_note` API to support a key (`C:maj` by default) that would sharp/flat notes in the right way.  Nobody's ever requested this functionality, but maybe it would be useful?

**Comment by bmcfee** [comment]
Pinging @kaustuvkanti on this one -- any interest in helping out with this?

**Comment by kaustuvkanti** [comment]
Totally, I'm in. Kindly advise.

    On Friday, 29 May, 2020, 8:01:40 pm GMT+4, Brian McFee <notifications@github.com> wrote:  
 
 


Pinging @kaustuvkanti on this one -- any interest in helping out with this?

**Comment by bmcfee** [comment]
> Totally, I'm in. Kindly advise.

Excellent, thanks!

The minimal version of this issue IMO is a unit converter for (hz/midi/note)<->svara.  From that, we can build out the display functionality (axis decorators for cqt and chroma representations).  My understanding is that such a converter would also need to provide a tonal center -- is that correct, and if so, how would it be specified?  I.e., what would a reasonable API and use-case look like from your perspective?

Apart from that, the technical points to be resolved are summarized in this comment: https://github.com/librosa/librosa/issues/641#issuecomment-344041427

**Comment by vrangasayee** [comment]
Thanks for the movement on this request. Yes we do need to provide a tonal
center. One way I can think of is to provide the tonal center as a concert
pitch ( my sa is at G#3 for example). The default could be sa <-> c3/c4?

Thanks
Vidya

**Comment by kaustuvkanti** [comment]
Yes, (hz/midi/note)<->svara is a rational abstraction. We should, however, be mindful of the differences in interpretation of the term "svara" in Hindustani versus Carnatic system. One other way is to plot a pitch salience histogram normalized with respect to the tonal center. This tonal hierarchy could be interesting too.. As of the tonal center, the Dunya CC collections have all tonic frequencies (tonal centers as you referred to) available through an API.
   On Sunday, 31 May, 2020, 1:20:46 am GMT+4, Vidya Rangasayee <notifications@github.com> wrote:  
 
 

Thanks for the movement on this request. Yes we do need to provide a tonal
center. One way I can think of is to provide the tonal center as a concert
pitch ( my sa is at G#3 for example). The default could be sa <-> c3/c4?

Thanks
Vidya

**Comment by bmcfee** [comment]
> Yes we do need to provide a tonal center. One way I can think of is to provide the tonal center as a concert pitch ( my sa is at G#3 for example). The default could be sa <-> c3/c4?

Right -- this now becomes a question of API: how should sa be represented?

- `note_to_svara`: As a 12-tet note?  Is octave important/necessary?
- `midi_to_svara`: As a midi number?
- `hz_to_svara`: As Hz?




> We should, however, be mindful of the differences in interpretation of the term "svara" in Hindustani versus Carnatic system.

Indeed.  This is another point that I haven't fully understood yet: should these be separate functions (`..._h`, `..._c`?), or should we have a keyword argument specifying which system to use?

Aside from naming / abbreviations, what else would need to change?  The number of octave subdivisions?  How much do we need to actually worry about equal temperament assumptions here?



> One other way is to plot a pitch salience histogram normalized with respect to the tonal center.

I think this idea is independent of the notation system being used -- one could just as easily apply that normalization to a pitch class profile.

For this issue, I'd like to keep things limited to unit conversion and display functionality.

---
As an aside, I've been thinking more that if we're going to add a tonal center option to this kind of unit conversion, we should do the same for the `_to_note()` functions.  This would allow us to be a bit more clever about how enharmonic equivalences are handled.  Relatedly, this could factor into chroma display by changing the locations of ticks to match up to prominent scale degrees.  Probably this should be a different issue, but some of the same thinking could be useful in how we plan out support for Indian systems.

**Comment by bmcfee** [comment]
Follow-up on the last comment, #1149 implements the key-conditional note spelling idea in our `midi_to_note` converter, which could be used as a template for the svara converter(s).

**Comment by kaustuvkanti** [comment]
I think hz_to_svara is the safest approach, as there are a considerable number of cases (not so much in the Dunya CC collection but overall) where the tuning deviates from the A-440 standard. However, note_to_svara can easily be adopted and octave is less important.

The functions ..._h and ..._c are indispensable, the definition of svara is very different in the two sub-genres. My idea of bringing in the histogram discussion was to learn the precise intonations by profiling the peaks. But for the case of Hindustani, assuming equal temperament is reasonable. The octave subdivisions are 12 in both, the nomenclature differs. The Carnatic system (left subfigure) uses 16 names for the 12-tet based on the raga context.
<img width="477" alt="Solfege_comparison" src="https://user-images.githubusercontent.com/4905391/84559392-43363580-ad4b-11ea-81e2-617de3447ff8.png">

The idea of chroma display sounds fantastic, I totally second your approach.

**Comment by bmcfee** [comment]
> I think hz_to_svara is the safest approach, as there are a considerable number of cases (not so much in the Dunya CC collection but overall) where the tuning deviates from the A-440 standard.

Not sure I'm following this.  Going from `hz->svara` necessarily involves quantization, and in both cases you describe, this can be accomplished in 12tet. This is essentially what `hz_to_midi` does, where the quantization is allowed to float (ie have fractional values like 60.5 for a half-sharp C𝄲).  If we need to quantize (to 12tet) anyway, I don't see why the core converter shouldn't be `midi_to_svara` instead of `hz_to_svara`?  Of course, once we have that core converter, we can build out helpers to handle the other cases `hz_to` and `note_to`.

This comes back to the question of what a natural representation of the base `sa` is.  In `midi_to_svara`, I would expect a (fractional) midi number; in `hz_to_svara`, I'd expect a frequency (to be converted to midi note).  Throwing western notes into this, except in the case of `note_to_svara` seems very complicated to me.


> The functions ..._h and ..._c are indispensable, the definition of svara is very different in the two sub-genres.

Is there a strong reason these should be separate functions rather than parameters to one function?  The names are going to be different, but it seems like there will be a lot redundant computation here if we split it into two methods. (Maybe I'm wrong on that.)



> But for the case of Hindustani, assuming equal temperament is reasonable.

Is it unreasonable for Carnatic?  (Sorry for the naive questions!)

I'd really like to avoid doing anything that depends on signal analysis here.  If any analysis does need to be done for analyzing a particular piece, I'd rather that happen elsewhere and its result be passed into the unit converter as a parameter.  (E.g., like `tuning`.)

> But for the case of Hindustani, assuming equal temperament is reasonable. The octave subdivisions are 12 in both, the nomenclature differs.

Are the subdivisions the same in both as well?  Or do we need to use something other than 12tet for Carnatic?


> The Carnatic system (left subfigure) uses 16 names for the 12-tet based on the raga context.

Right.  This to me seems structurally similar to note spelling in western keys, which we recently implemented #1149 .  Could you say more (or point to a good ref) for how raga context can be encoded?  In `midi_to_note`, the `key` parameter corresponds to one of 24 keys (12 chromatic notes * {major, minor}), though we do a bit of regular expression magic to allow a more expressive interface.  Is something like that possible here?

**Comment by sertansenturk** [comment]
Hi everyone,

Joining the conversation upon @kaustuvkanti's request. **Disclaimer:** I do not consider myself as I culturally/musicologically understand Indian classical music traditions. I'm basically writing as a researcher who did some computational work on the topic. I will try to be as clear as possible but please let me know if things do not make sense. Also @kaustuvkanti please don't hesitate to correct any mistake I make.

A few years ago, we published a paper in SMC, where we tried to capture the pitch distributions for individual svaras per raga in Carnatic Music ([link](https://sertansenturk.com/uploads/publications/senturk2016noteModel_smc.pdf)). We achieved it by aligning the music scores of the compositions to the relevant performances.

**In short:** We can align the svaras so talking about a symbolic representation is relevant. However, interpreting them as a single/centre frequency etc. is maybe reductionist. Below I paste a Figure from the paper, which shows how drastically the same svara deviates from the just-intonation temperament in two different ragas.

![Screen Shot 2020-06-18 at 16 17 58](https://user-images.githubusercontent.com/4540641/85039158-59932580-b17f-11ea-9263-4e5d80f0c65e.png)

**Comment by bmcfee** [comment]
Thanks for this @sertansenturk -- I'm not sure I understand your post though.

We expect the f0 for different svaras to change per recording, which is why we need to provide the reference sa frequency in the conversion.  The issue we're trying to address here is how to annotate frequencies once the sa and raga are given.  The main use-case of this is in display/plotting (just like `hz_to_note`), but it would also be useful to have as a utility a function.  If `sa` changes over time in a track (and you have a way to estimate that), this converter could be used to normalize deviations like the ones in your plot.

The main blocker on this issue right now is how to construct the mapping from raga to note spelling, as well as the questions in my previous post that I'm still waiting for response on.

**Comment by bmcfee** [comment]
> The main blocker on this issue right now is how to construct the mapping from raga to note spelling, as well as the questions in my previous post that I'm still waiting for response on.

I did a bit more reading on the topic, and have a slightly better grasp of what the issues are now.  @kaustuvkanti would it be reasonable, for Carnatic support, to only implement melakarta? This could be compactly implemented by a look-up table, similar to (but much larger than) what we do for major/minor presently.  It seems to me that this would cover most of the use cases, and if a user's in janya land, it's on them to adapt.

The other question I have is about notation for pitch height: how standard is the over-dot/under-dot convention?  And what would we do if wandering outside the three-octave range?  In SPN, we just tack an octave number onto the end of the note, but this would be confusing if we already have numbers as part of the note name (eg N1/N2).

**Comment by bmcfee** [comment]
Took a cut at prototyping some of this today, and I think it's mostly going to work out.  The basic Hindustani conversion code is there, Carnatic isn't implemented yet but I've got a clear idea for how it will go now.

Some updates:

- Decided to go with separate function families for Hindustani and Carnatic.  There's some similarity here, but functionally there's very little reusable computation.
- The core computations are implemented as `midi_to_svara_h` and `..._c`.  The following helpers will exist:
  - `hz_to_svara_*`, `note_to_svara_*`.  These all run through the midi converter, and allow you to specify `Sa` in whatever format you're converting from. Eg, `hz_to_svara_h(frequencies, Sa=261.62)`, `note_to_svara_h(notes, Sa='C2')`, `midi_to_svara_h(midis, Sa=60)`.
  - This implicitly requires 12-tet for note spacing, but Sa need not be on the 440/12tet grid. Fractional midi numbers are used to keep the conversion 1-1.
  - The 12-tet note spacing is for sure a compromise, but it's the simplest way to get consistency with the rest of the package.  Maybe in the future we can add a `just=True` flag to use just intonation if there's sufficient demand.
- Added a `thaat_to_degrees()` function, similar to `key_to_degrees`, which (for Hindustani) returns the indices of notes in the thaat starting from Sa.  This can be used to specify the tick locations of a chroma plot, for example.
- Converter uses some unicode hackery to get (optional) +-1 octave decorations (relative to the home octave) via over- and under-dots.  If `unicode=False` is given, these are converted to `'` and `,`.
- Conversion can be `abbr=True` to get (`SrRgGmMPdDnN`) or `abbr=False` to get `Sa, re, Re, ga, Ga, ...`.  Octave decorations are applied to the first character (unicode: `Ṡ`, `Ṣa`) or as suffices (non-unicode, `S'`, `Sa,`).


Call signature for the core converter looks like:
```python
midi_to_svara_h(midi, Sa=60, abbr=True, octave=True, unicode=True)
```
---

The Carnatic converter will look similar, but use the numbering convention for svara, eg `R₁` instead of `r`.  (The same optional unicode conventions will apply for the typesetting here, `unicode=False` would give `R1` instead.)

It will also accept a parameter `mela=` which will operate similar to the `key=` parameter in `midi_to_note` to determine svara spelling.  This will be supported by an index structure which maps each raga to its index (0-71) and then does the appropriate modular arithmetic to determine the positions and spellings of M, R, G, D, and N.  A similar helper function `mela_to_degrees` will provide the positioning information.
---
On the `display` side, once the converters are stable, we'll add new decorators `cqt_svara_*` and `chroma_svara_*`, implemented by analogous classes to `NoteFormatter` and `ChromaFormatter`.  Ticker locations will be determined by the appropriate `_to_degrees` function, supported by either `thaat=` or `mela=` parameters as appropriate.

---
Remaining questions:

1. Octave decorations are not applied outside the +-1 octave range.  This will make a full inversion (`svara_h_to_midi`) impossible.  Perhaps we need to extend the notation to allow for arbitrary octave offsets?  Or maybe just not implement the inverse transforms?
2. Should the `_to_degrees` helpers be absolute or relative?  In `key_to_degrees` they're absolute in the sense that they're always offsets from C.  This is helpful for things like chroma viz because it anchors the representation in a way that's independent of the recording (bin 0 is always measuring C, bin 1 is measuring C#, etc).  As I've currently implemented `thaat_to_degrees`, it's relative to Sa (whatever that happens to be), which is probably the right thing to do but will make visualization awkward (at least for chroma).  I don't see any clean way to make it absolute though, so perhaps we'll just have to live with it this way.  But any input would be appreciated!
